src/client/csfle/client_encryption.rs

```diff
@@ -0,0 +1,354 @@
+};
+use serde::Serialize;
+use typed_builder::TypedBuilder;
+
+use super::{
+    options::{KmsProviders, KmsProvidersTlsOptions},
+    state_machine::CryptExecutor,
+};
+
+/// A handle to the key vault.  Used to create data encryption keys, and to explicitly encrypt and
+/// decrypt values when auto-encryption is not an option.
+pub struct ClientEncryption {
+    crypt: Crypt,
+    exec: CryptExecutor,
+    key_vault: Collection<RawDocumentBuf>,
+}
+
+impl ClientEncryption {
+    /// Create a new key vault handle with the given options.
+    pub fn new(options: ClientEncryptionOptions) -> Result<Self> {
+        let crypt = Crypt::builder().build()?;
+        let exec = CryptExecutor::new(
+            options.key_vault_client.weak(),
+            options.key_vault_namespace.clone(),
+            None,
+            None,
+            options.tls_options,
+        )?;
+        let key_vault = options
+            .key_vault_client
+            .database(&options.key_vault_namespace.db)
+            .collection_with_options(
+                &options.key_vault_namespace.coll,
+                CollectionOptions::builder()
+                    .write_concern(WriteConcern::MAJORITY)
+                    .read_concern(ReadConcern::MAJORITY)
+                    .build(),
+            );
+        Ok(Self {
+            crypt,
+            exec,
+            key_vault,
+        })
+    }
+
+    /// Creates a new key document and inserts into the key vault collection.
+    /// Returns the _id of the created document as a UUID (BSON binary subtype 0x04).
+    pub async fn create_data_key(
+        &self,
+        kms_provider: KmsProvider,
+        opts: DataKeyOptions,
+    ) -> Result<Binary> {
+        let ctx = self.create_data_key_ctx(kms_provider, opts)?;
+        let data_key = self.exec.run_ctx(ctx, None).await?;
+        self.key_vault.insert_one(&data_key, None).await?;
+        let bin_ref = data_key
+            .get_binary("_id")
+            .map_err(|e| Error::internal(format!("invalid data key id: {}", e)))?;
+        Ok(bin_owned(bin_ref))
+    }
+
+    fn create_data_key_ctx(&self, kms_provider: KmsProvider, opts: DataKeyOptions) -> Result<Ctx> {
+        let mut builder = self.crypt.ctx_builder();
+        let mut key_doc = doc! { "provider": kms_provider.name() };
+        if !matches!(opts.master_key, MasterKey::Local) {
+            let master_doc = bson::to_document(&opts.master_key)?;
+            key_doc.extend(master_doc);
+        }
+        builder = builder.key_encryption_key(&key_doc)?;
+        if let Some(alt_names) = opts.key_alt_names {
+            for name in alt_names {
+                builder = builder.key_alt_name(&name)?;
+            }
+        }
+        if let Some(material) = opts.key_material {
+            builder = builder.key_material(&material)?;
+        }
+        Ok(builder.build_datakey()?)
+    }
+
+    // pub async fn rewrap_many_data_key(&self, _filter: Document, _opts: RewrapManyDataKeyOptions)
```

Comment:
Are these commented sections here and below intentionally left in?

Reply:
Leaving them sound good, figured that was the case but just wanted to make sure.